File: eco-core/core-plugin/src/main/java/com/willfp/stattrackers/StatTrackersPlugin.java
package com.willfp.stattrackers;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.command.impl.PluginCommand;
import com.willfp.eco.core.display.DisplayModule;
import com.willfp.stattrackers.commands.CommandActivestat;
import com.willfp.stattrackers.commands.CommandGive;
import com.willfp.stattrackers.commands.CommandReload;
import com.willfp.stattrackers.commands.CommandStatTrackers;
import com.willfp.stattrackers.display.StatTrackersDisplay;
import com.willfp.stattrackers.stats.Stats;
import com.willfp.stattrackers.tracker.TrackerListener;
import lombok.Getter;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;

@SuppressWarnings("unused")
public class StatTrackersPlugin extends EcoPlugin {
    /**
     * Instance of the plugin.
     */
    @Getter
    private static StatTrackersPlugin instance;

    /**
     * Internal constructor called by bukkit on plugin load.
     */
    public StatTrackersPlugin() {
        super(623, 10261, "&d", true);
        instance = this;
    }

    @Override
    protected void handleEnable() {
        this.getLogger().info(Stats.values().size() + " Stats Loaded");
    }

    @Override
    protected void handleReload() {
        Stats.values().forEach(stat -> {
            HandlerList.unregisterAll(stat);

            this.getScheduler().runLater(() -> {
                this.getEventManager().registerListener(stat);
            }, 1);
        });
    }

    @Override
    protected List<PluginCommand> loadPluginCommands() {
        return Arrays.asList(
                new CommandStatTrackers(this)
        );
    }

    @Override
    protected List<Listener> loadListeners() {
        return Arrays.asList(
                new TrackerListener(this)
        );
    }

    @Override
    protected @Nullable DisplayModule createDisplayModule() {
        return new StatTrackersDisplay(this);
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/stattrackers/StatTrackersPlugin.kt
package com.willfp.stattrackers

import com.willfp.eco.core.bstats.EcoMetricsChart
import com.willfp.eco.core.command.impl.PluginCommand
import com.willfp.eco.core.display.DisplayModule
import com.willfp.libreforge.loader.LibreforgePlugin
import com.willfp.libreforge.loader.configs.ConfigCategory
import com.willfp.stattrackers.commands.CommandStatTrackers
import com.willfp.stattrackers.config.TargetsYml
import com.willfp.stattrackers.display.StatTrackersDisplay
import com.willfp.stattrackers.stats.StatTargets
import com.willfp.stattrackers.stats.StatTrackersGUI
import com.willfp.stattrackers.stats.Stats
import com.willfp.stattrackers.util.DiscoverRecipeListener
import org.bukkit.event.Listener

internal lateinit var plugin: StatTrackersPlugin
    private set

class StatTrackersPlugin : LibreforgePlugin() {
    val targetsYml = TargetsYml(this)

    init {
        plugin = this
    }

    override fun handleReload() {
        StatTargets.update()
        StatTrackersGUI.update()
    }

    override fun loadConfigCategories(): List<ConfigCategory> {
        return listOf(
            Stats
        )
    }

    override fun loadPluginCommands(): List<PluginCommand> {
        return listOf(
            CommandStatTrackers
        )
    }

    override fun loadListeners(): List<Listener> {
        return listOf(
            DiscoverRecipeListener
        )
    }

    @Suppress("OVERRIDE_DEPRECATION")
    override fun createDisplayModule(): DisplayModule {
        return StatTrackersDisplay
    }

    override fun getCustomCharts() = listOf(
        EcoMetricsChart.SingleLine("total_stats") { Stats.values().size },
        EcoMetricsChart.SingleLine("total_stat_targets") { StatTargets.values().size },
        EcoMetricsChart.SimplePie("discover_recipes") {
            if (configYml.getBool("discover-recipes")) "enabled" else "disabled"
        },
        EcoMetricsChart.SimplePie("display_at_top") {
            if (configYml.getBool("display-at-top")) "top" else "bottom"
        }
    )

}
